backend/src/main/java/fk/prof/backend/aggregator/ProfileWorkInfo.java

```diff
@@ -2,80 +2,112 @@
-    this.associatedWorkTypes.add(workType);
-  }
+  private final HashObjIntMap<Recorder.WorkType> workTypeSamples = HashObjIntMaps.newUpdatableMap();
 
-  /**
-   * Returns unmodifiable view of trace name -> trace coverage data associated with a work id
-   * @return lookup map for trace coverage using trace name
-   */
-  public Map<String, Integer> getTraceCoverages() {
-    return this.traceCoveragesUnmodifiableView;
+  public void updateWSESpecificDetails(Recorder.Wse wse) {
+    for (Recorder.TraceContext trace : wse.getIndexedData().getTraceCtxList()) {
+      traceCoverages.put(trace.getTraceName(), trace.getCoveragePct());
+    }
+    workTypeSamples.put(wse.getWType(), workTypeSamples.getOrDefault(wse.getWType(), 0) + getSampleCount(wse));
   }
 
-  /**
-   * Returns unmodifiable view of all work types associated with the workId
-   * @return set of associated work types
-   */
-  public Set<Recorder.WorkType> getAssociatedWorkTypes() {
-    return this.associatedWorkTypesUnmodifiableView;
-  }
-
-  public void setStatus(AggregationStatus aggregationStatus) {
-    status = aggregationStatus;
-  }
-
-  public AggregationStatus getStatus() {
-    return status;
+  public AggregationState startProfile(int recorderVersion, LocalDateTime startedAt) {
+    if(!processStateEvent(AggregationStateEvent.START_PROFILE)) {
+      throw new IllegalStateException(String.format("Invalid event %s for current state %s",
+          AggregationStateEvent.START_PROFILE, state));
+    }
+    this.recorderVersion = recorderVersion;
+    this.startedAt = startedAt;
+    return state;
   }
 
-  public LocalDateTime getStartedAt() {
-    return startedAt;
+  public AggregationState completeProfile() {
+    if(!processStateEvent(AggregationStateEvent.COMPLETE_PROFILE)) {
+      throw new IllegalStateException(String.format("Invalid event %s for current state %s",
+          AggregationStateEvent.COMPLETE_PROFILE, state));
+    }
+    this.endedAt = LocalDateTime.now(Clock.systemUTC());
+    return state;
   }
 
-  public void setStartedAt(LocalDateTime startedAt) {
-    this.startedAt = startedAt;
+  public AggregationState abandonProfile() {
+    if(!processStateEvent(AggregationStateEvent.ABANDON_PROFILE)) {
+      throw new IllegalStateException(String.format("Invalid event %s for current state %s",
+          AggregationStateEvent.ABANDON_PROFILE, state));
+    }
+    this.endedAt = LocalDateTime.now(Clock.systemUTC());
+    return state;
   }
 
-  public LocalDateTime getEndedAt() {
-    return endedAt;
+  public AggregationState abortProfile() {
+    if(!processStateEvent(AggregationStateEvent.ABORT_PROFILE)) {
+      throw new IllegalStateException(String.format("Invalid event %s for current state %s",
+          AggregationStateEvent.ABORT_PROFILE, state));
+    }
+    return state;
   }
 
-  public void setEndedAt(LocalDateTime endedAt) {
-    this.endedAt = endedAt;
+  private boolean processStateEvent(AggregationStateEvent stateEvent) {
```

Comment:
The methods of this class are not thread-safe (documented at class start). Since this is an object specific to work-id, it should always be accessed from a single thread (since we also ensure same work-id can not be processed concurrently)
If this is only for single threaded access, is atomicity a hard requirement here?